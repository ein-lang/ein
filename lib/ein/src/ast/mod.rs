mod alternative;
mod application;
mod boolean;
mod builtin_interface;
mod case;
mod definition;
mod export;
mod expression;
mod function_definition;
mod if_;
mod import;
mod let_;
mod list;
mod list_case;
mod list_element;
mod module;
mod module_interface;
mod none;
mod number;
mod operation;
mod operator;
mod record_construction;
mod record_element_operation;
mod record_update;
mod string;
mod type_coercion;
mod type_definition;
mod unresolved_import;
mod unresolved_module;
mod variable;
mod variable_definition;

pub use alternative::*;
pub use application::*;
pub use boolean::*;
pub use builtin_interface::*;
pub use case::*;
pub use definition::*;
pub use export::*;
pub use expression::*;
pub use function_definition::*;
pub use if_::*;
pub use import::*;
pub use let_::*;
pub use list::*;
pub use list_case::*;
pub use list_element::*;
pub use module::*;
pub use module_interface::*;
pub use none::*;
pub use number::*;
pub use operation::*;
pub use operator::*;
pub use record_construction::*;
pub use record_element_operation::*;
pub use record_update::*;
pub use string::*;
pub use type_coercion::*;
pub use type_definition::*;
pub use unresolved_import::*;
pub use unresolved_module::*;
pub use variable::*;
pub use variable_definition::*;
