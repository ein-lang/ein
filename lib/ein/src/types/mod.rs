mod anonymous_record;
mod boolean;
mod function;
mod none;
mod number;
mod record;
mod reference;
mod type_;
mod unknown;
mod variable;

pub use anonymous_record::*;
pub use boolean::*;
pub use function::*;
pub use none::*;
pub use number::*;
pub use record::*;
pub use reference::*;
pub use type_::*;
pub use unknown::*;
pub use variable::*;
