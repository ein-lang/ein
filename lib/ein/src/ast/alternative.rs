use super::expression::Expression;
use super::pattern::Pattern;
use crate::types::Type;
use std::collections::HashMap;

#[derive(Clone, Debug, PartialEq)]
pub struct Alternative {
    pattern: Pattern,
    expression: Expression,
}

impl Alternative {
    pub fn new(pattern: impl Into<Pattern>, expression: impl Into<Expression>) -> Self {
        Self {
            pattern: pattern.into(),
            expression: expression.into(),
        }
    }

    pub fn pattern(&self) -> &Pattern {
        &self.pattern
    }

    pub fn expression(&self) -> &Expression {
        &self.expression
    }

    pub fn convert_expressions(&self, convert: &mut impl FnMut(&Expression) -> Expression) -> Self {
        Self::new(
            self.pattern.clone(),
            self.expression.convert_expressions(convert),
        )
    }

    pub fn convert_types(&self, convert: &mut impl FnMut(&Type) -> Type) -> Self {
        Self::new(self.pattern.clone(), self.expression.convert_types(convert))
    }

    pub fn resolve_reference_types(&self, environment: &HashMap<String, Type>) -> Self {
        Self::new(
            self.pattern.clone(),
            self.expression.resolve_reference_types(environment),
        )
    }
}
