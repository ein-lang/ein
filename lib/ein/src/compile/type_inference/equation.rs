use crate::types::*;

// TODO Consider renaming as it is not symmetric.
#[derive(Clone, Debug, Eq, Hash, PartialEq)]
pub struct Equation {
    lhs: Type,
    rhs: Type,
}

impl Equation {
    pub fn new(lhs: impl Into<Type>, rhs: impl Into<Type>) -> Self {
        Self {
            lhs: lhs.into(),
            rhs: rhs.into(),
        }
    }

    pub fn lhs(&self) -> &Type {
        &self.lhs
    }

    pub fn rhs(&self) -> &Type {
        &self.rhs
    }
}
