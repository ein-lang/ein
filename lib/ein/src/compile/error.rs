use crate::ast;
use crate::debug::*;
use crate::types;
use std::error::Error;
use std::fmt::Display;
use std::rc::Rc;

#[derive(Debug, PartialEq)]
pub enum CompileError {
    CircularInitialization,
    ExportedNameNotFound { name: String },
    MixedDefinitionsInLet(Rc<SourceInformation>),
    SsfAnalysis(ssf::AnalysisError),
    SsfCompile(ssf_llvm::CompileError),
    TypeNotFound(types::Reference),
    TypesNotMatched(Rc<SourceInformation>, Rc<SourceInformation>),
    VariableNotFound(ast::Variable),
}

impl Display for CompileError {
    fn fmt(&self, formatter: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
        match self {
            Self::CircularInitialization => {
                write!(formatter, "circular variable initialization detected",)
            }
            Self::ExportedNameNotFound { name } => {
                write!(formatter, "exported name \"{}\" not found", name)
            }
            Self::MixedDefinitionsInLet(source_information) => write!(
                formatter,
                "cannot mix function and value definitions in a let expression\n{}",
                source_information
            ),
            Self::SsfAnalysis(error) => write!(formatter, "{}", error),
            Self::SsfCompile(error) => write!(formatter, "{}", error),
            Self::TypeNotFound(reference) => write!(
                formatter,
                "type \"{}\" not found\n{}",
                reference.name(),
                reference.source_information()
            ),
            Self::TypesNotMatched(lhs_source_information, rhs_source_information) => write!(
                formatter,
                "types not matched\n{}\n{}",
                lhs_source_information, rhs_source_information
            ),
            Self::VariableNotFound(variable) => write!(
                formatter,
                "variable \"{}\" not found\n{}",
                variable.name(),
                variable.source_information()
            ),
        }
    }
}

impl Error for CompileError {}

impl From<ssf::AnalysisError> for CompileError {
    fn from(error: ssf::AnalysisError) -> Self {
        match error {
            ssf::AnalysisError::CircularInitialization => Self::CircularInitialization,
            _ => Self::SsfAnalysis(error),
        }
    }
}

impl From<ssf_llvm::CompileError> for CompileError {
    fn from(error: ssf_llvm::CompileError) -> Self {
        Self::SsfCompile(error)
    }
}
