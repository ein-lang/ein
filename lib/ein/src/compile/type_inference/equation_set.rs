use super::equation::Equation;
use std::collections::HashSet;

#[derive(Debug)]
pub struct EquationSet {
    cache: HashSet<Equation>,
    equations: Vec<Equation>,
}

impl EquationSet {
    pub fn new() -> Self {
        Self {
            cache: HashSet::new(),
            equations: vec![],
        }
    }

    pub fn add(&mut self, equation: Equation) {
        if self.cache.contains(&equation) {
            return;
        }

        self.cache.insert(equation.clone());
        self.equations.push(equation);
    }

    pub fn remove(&mut self) -> Option<Equation> {
        self.equations.pop()
    }

    pub fn iter_mut(&mut self) -> impl IntoIterator<Item = &mut Equation> {
        self.equations.iter_mut()
    }
}
