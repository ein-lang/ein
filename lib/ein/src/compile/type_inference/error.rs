use crate::debug::SourceInformation;
use crate::types;
use std::error::Error;
use std::fmt::Display;
use std::rc::Rc;

#[derive(Clone, Debug, PartialEq)]
pub enum TypeInferenceError {
    TypesNotMatched(Rc<SourceInformation>, Rc<SourceInformation>),
    TypeNotFound { reference: types::Reference },
    VariableNotFound(String, Rc<SourceInformation>),
}

impl TypeInferenceError {}

impl Display for TypeInferenceError {
    fn fmt(&self, formatter: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
        match self {
            TypeInferenceError::TypesNotMatched(lhs_source_information, rhs_source_information) => {
                write!(
                    formatter,
                    "types not matched\n{}\n{}",
                    lhs_source_information, rhs_source_information
                )
            }
            TypeInferenceError::TypeNotFound { reference } => write!(
                formatter,
                "type \"{}\" not found\n{}",
                reference.name(),
                reference.source_information()
            ),
            TypeInferenceError::VariableNotFound(name, source_information) => write!(
                formatter,
                "variable \"{}\" not found\n{}",
                name, source_information
            ),
        }
    }
}

impl Error for TypeInferenceError {}
