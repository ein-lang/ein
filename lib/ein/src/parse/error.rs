use super::input::Input;
use crate::debug::*;
use std::error::Error;
use std::fmt::Display;

#[derive(Clone, Debug, PartialEq)]
pub struct ParseError {
    source_information: Box<SourceInformation>,
}

impl ParseError {
    pub fn new(input: &Input) -> Self {
        ParseError {
            source_information: SourceInformation::new(
                input.source().name(),
                input.location(),
                input.line(),
            )
            .into(),
        }
    }
}

impl Display for ParseError {
    fn fmt(&self, formatter: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
        write!(
            formatter,
            "ParseError: Failed to parse\n{}",
            self.source_information,
        )
    }
}

impl Error for ParseError {
    fn description(&self) -> &str {
        "ParseError"
    }
}
