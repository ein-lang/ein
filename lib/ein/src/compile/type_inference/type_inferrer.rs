use super::equation::*;
use super::equation_set::EquationSet;
use super::error::*;
use crate::ast::*;
use crate::types::{self, Type};
use std::collections::*;

#[derive(Debug)]
pub struct TypeInferrer {
    environment: HashMap<String, Type>,
    equation_set: EquationSet,
}

impl TypeInferrer {
    pub fn new() -> Self {
        Self {
            environment: HashMap::new(),
            equation_set: EquationSet::new(),
        }
    }

    pub fn infer(&mut self, module: &Module) -> Result<Module, TypeInferenceError> {
        for imported_module in module.imported_modules() {
            self.environment.extend(
                imported_module.types().iter().map(|(name, type_)| {
                    (imported_module.path().qualify_name(name), type_.clone())
                }),
            );
        }

        for type_definition in module.type_definitions() {
            self.environment.insert(
                type_definition.name().into(),
                type_definition.type_().clone(),
            );
        }

        self.collect_equations(module)?;

        let substitutions = self.reduce_equations()?;

        Ok(module.convert_types(&mut |type_| {
            if let Type::Variable(variable) = type_ {
                substitutions[&variable.id()].clone()
            } else {
                type_.clone()
            }
        }))
    }

    fn collect_equations(&mut self, module: &Module) -> Result<(), TypeInferenceError> {
        let mut variables = HashMap::<String, Type>::new();

        for imported_module in module.imported_modules() {
            for (name, type_) in imported_module.variables() {
                variables.insert(imported_module.path().qualify_name(name), type_.clone());
            }
        }

        for definition in module.definitions() {
            match definition {
                Definition::FunctionDefinition(function_definition) => {
                    variables.insert(
                        function_definition.name().into(),
                        function_definition.type_().clone(),
                    );
                }
                Definition::ValueDefinition(value_definition) => {
                    variables.insert(
                        value_definition.name().into(),
                        value_definition.type_().clone(),
                    );
                }
            }
        }

        for definition in module.definitions() {
            match definition {
                Definition::FunctionDefinition(function_definition) => {
                    self.infer_function_definition(function_definition, &variables)?;
                }
                Definition::ValueDefinition(value_definition) => {
                    self.infer_value_definition(value_definition, &variables)?;
                }
            };
        }

        Ok(())
    }

    fn infer_function_definition(
        &mut self,
        function_definition: &FunctionDefinition,
        variables: &HashMap<String, Type>,
    ) -> Result<(), TypeInferenceError> {
        let source_information = function_definition.source_information();
        let mut variables = variables.clone();
        let mut type_ = function_definition.type_().clone();

        for argument_name in function_definition.arguments() {
            let argument_type: Type = types::Variable::new(source_information.clone()).into();
            let result_type: Type = types::Variable::new(source_information.clone()).into();

            self.equation_set.add(Equation::new(
                type_,
                types::Function::new(
                    argument_type.clone(),
                    result_type.clone(),
                    source_information.clone(),
                ),
            ));

            variables.insert(argument_name.into(), argument_type);

            type_ = result_type;
        }

        let body_type = self.infer_expression(function_definition.body(), &variables)?;
        self.equation_set.add(Equation::new(body_type, type_));

        Ok(())
    }

    fn infer_value_definition(
        &mut self,
        value_definition: &ValueDefinition,
        variables: &HashMap<String, Type>,
    ) -> Result<(), TypeInferenceError> {
        let type_ = self.infer_expression(value_definition.body(), &variables)?;

        self.equation_set
            .add(Equation::new(type_, value_definition.type_().clone()));

        Ok(())
    }

    fn infer_expression(
        &mut self,
        expression: &Expression,
        variables: &HashMap<String, Type>,
    ) -> Result<Type, TypeInferenceError> {
        match expression {
            Expression::Application(application) => {
                let function = self.infer_expression(application.function(), variables)?;
                let argument = self.infer_expression(application.argument(), variables)?;
                let result: Type =
                    types::Variable::new(application.source_information().clone()).into();

                self.equation_set.add(Equation::new(
                    function,
                    types::Function::new(
                        argument,
                        result.clone(),
                        application.source_information().clone(),
                    ),
                ));

                Ok(result)
            }
            Expression::Boolean(boolean) => {
                Ok(types::Boolean::new(boolean.source_information().clone()).into())
            }
            Expression::If(if_) => {
                let condition = self.infer_expression(if_.condition(), variables)?;
                self.equation_set.add(Equation::new(
                    condition,
                    types::Boolean::new(if_.source_information().clone()),
                ));

                let then = self.infer_expression(if_.then(), variables)?;
                let else_ = self.infer_expression(if_.else_(), variables)?;

                self.equation_set.add(Equation::new(then.clone(), else_));

                Ok(then)
            }
            Expression::Let(let_) => {
                let mut variables = variables.clone();

                // Because the language does not have let-rec
                // expression like OCaml, we need to guess if the
                // let expression is recursive or not to generate
                // proper type equations.
                let functions_included =
                    let_.definitions()
                        .iter()
                        .any(|definition| match definition {
                            Definition::FunctionDefinition(_) => true,
                            Definition::ValueDefinition(value_definition) => {
                                if let Type::Function(_) = value_definition.type_() {
                                    true
                                } else {
                                    false
                                }
                            }
                        });

                for definition in let_.definitions() {
                    match definition {
                        Definition::FunctionDefinition(function_definition) => {
                            variables.insert(
                                function_definition.name().into(),
                                function_definition.type_().clone(),
                            );
                        }
                        Definition::ValueDefinition(value_definition) => {
                            if functions_included {
                                variables.insert(
                                    value_definition.name().into(),
                                    value_definition.type_().clone(),
                                );
                            }
                        }
                    }
                }

                for definition in let_.definitions() {
                    match definition {
                        Definition::FunctionDefinition(function_definition) => {
                            self.infer_function_definition(function_definition, &variables)?;
                        }
                        Definition::ValueDefinition(value_definition) => {
                            self.infer_value_definition(value_definition, &variables)?;

                            variables.insert(
                                value_definition.name().into(),
                                value_definition.type_().clone(),
                            );
                        }
                    }
                }

                self.infer_expression(let_.expression(), &variables)
            }
            Expression::None(none) => {
                Ok(types::None::new(none.source_information().clone()).into())
            }
            Expression::Number(number) => {
                Ok(types::Number::new(number.source_information().clone()).into())
            }
            Expression::Operation(operation) => {
                let type_: Type = types::Number::new(operation.source_information().clone()).into();

                let lhs = self.infer_expression(operation.lhs(), variables)?;
                self.equation_set.add(Equation::new(lhs, type_.clone()));
                let rhs = self.infer_expression(operation.rhs(), variables)?;
                self.equation_set.add(Equation::new(rhs, type_.clone()));

                Ok(type_)
            }
            Expression::Record(record) => {
                let type_ = types::AnonymousRecord::new(
                    record
                        .elements()
                        .iter()
                        .map(|(key, expression)| {
                            Ok((key.clone(), self.infer_expression(expression, variables)?))
                        })
                        .collect::<Result<_, _>>()?,
                    record.source_information().clone(),
                );

                self.equation_set
                    .add(Equation::new(record.type_().clone(), type_));

                Ok(record.type_().clone())
            }
            Expression::RecordElementOperation(operation) => {
                let source_information = operation.source_information();
                let argument = self.infer_expression(operation.argument(), variables)?;
                let result = types::Variable::new(source_information.clone());

                self.equation_set.add(Equation::new(
                    types::AnonymousRecord::new(
                        vec![(operation.key().into(), result.clone().into())]
                            .into_iter()
                            .collect(),
                        source_information.clone(),
                    ),
                    argument.clone(),
                ));
                self.equation_set
                    .add(Equation::new(operation.type_().clone(), argument));

                Ok(result.into())
            }
            Expression::RecordElementOperator(_) => unreachable!(),
            Expression::Variable(variable) => {
                variables.get(variable.name()).cloned().ok_or_else(|| {
                    TypeInferenceError::VariableNotFound(
                        variable.name().into(),
                        variable.source_information().clone(),
                    )
                })
            }
        }
    }

    fn reduce_equations(&mut self) -> Result<HashMap<usize, Type>, TypeInferenceError> {
        let mut substitutions = HashMap::<usize, Type>::new();

        while let Some(equation) = self.equation_set.remove() {
            match (equation.lhs(), equation.rhs()) {
                (Type::Variable(variable), rhs) => {
                    if let Type::Variable(other_variable) = rhs {
                        if variable.id() == other_variable.id() {
                            continue;
                        }
                    }

                    self.substitute_variable(variable, rhs, &mut substitutions);
                }
                (lhs, Type::Variable(variable)) => {
                    self.substitute_variable(variable, lhs, &mut substitutions);
                }
                (Type::Reference(reference), other) => self.equation_set.add(Equation::new(
                    self.environment
                        .get(reference.name())
                        .ok_or_else(|| TypeInferenceError::TypeNotFound {
                            reference: reference.clone(),
                        })?
                        .clone(),
                    other.clone(),
                )),
                (one, Type::Reference(reference)) => self.equation_set.add(Equation::new(
                    one.clone(),
                    self.environment
                        .get(reference.name())
                        .ok_or_else(|| TypeInferenceError::TypeNotFound {
                            reference: reference.clone(),
                        })?
                        .clone(),
                )),
                (Type::Function(one), Type::Function(other)) => {
                    self.equation_set.add(Equation::new(
                        one.argument().clone(),
                        other.argument().clone(),
                    ));

                    // Handle contravariance.
                    self.equation_set
                        .add(Equation::new(other.result().clone(), one.result().clone()));
                }
                (Type::Boolean(_), Type::Boolean(_)) => {}
                (Type::None(_), Type::None(_)) => {}
                (Type::Number(_), Type::Number(_)) => {}
                (Type::Record(one), Type::Record(other)) => {
                    if one.name() != other.name() {
                        return Err(TypeInferenceError::TypesNotMatched(
                            one.source_information().clone(),
                            other.source_information().clone(),
                        ));
                    };
                }
                (Type::Record(record), Type::AnonymousRecord(anonymous_record)) => {
                    let error = TypeInferenceError::TypesNotMatched(
                        record.source_information().clone(),
                        anonymous_record.source_information().clone(),
                    );

                    if record.elements().len() != anonymous_record.elements().len() {
                        return Err(error);
                    }

                    for (key, type_) in record.elements() {
                        self.equation_set.add(Equation::new(
                            type_.clone(),
                            anonymous_record
                                .elements()
                                .get(key)
                                .ok_or_else(|| error.clone())?
                                .clone(),
                        ));
                    }
                }
                (Type::AnonymousRecord(anonymous_record), Type::Record(record)) => {
                    let error = TypeInferenceError::TypesNotMatched(
                        anonymous_record.source_information().clone(),
                        record.source_information().clone(),
                    );

                    if anonymous_record.elements().len() > record.elements().len() {
                        return Err(error);
                    }

                    for (key, type_) in anonymous_record.elements() {
                        self.equation_set.add(Equation::new(
                            type_.clone(),
                            record
                                .elements()
                                .get(key)
                                .ok_or_else(|| error.clone())?
                                .clone(),
                        ));
                    }
                }
                (lhs, rhs) => {
                    return Err(TypeInferenceError::TypesNotMatched(
                        lhs.source_information().clone(),
                        rhs.source_information().clone(),
                    ))
                }
            }
        }

        Ok(substitutions)
    }

    fn substitute_variable(
        &mut self,
        variable: &types::Variable,
        type_: &Type,
        substitutions: &mut HashMap<usize, Type>,
    ) {
        for (_, substituted_type) in substitutions.iter_mut() {
            *substituted_type = substituted_type
                .clone()
                .substitute_variable(variable, type_);
        }

        for equation in self.equation_set.iter_mut() {
            *equation = Equation::new(
                equation.lhs().substitute_variable(variable, type_),
                equation.rhs().substitute_variable(variable, type_),
            )
        }

        substitutions.insert(variable.id(), type_.clone());
    }
}
