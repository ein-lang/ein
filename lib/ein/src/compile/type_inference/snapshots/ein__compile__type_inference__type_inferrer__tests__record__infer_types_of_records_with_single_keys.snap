---
source: lib/ein/src/compile/type_inference/type_inferrer.rs
expression: "infer_types(&Module::from_definitions_and_type_definitions(vec![TypeDefinition\n                                                                ::\n                                                                new(\"Foo\",\n                                                                    record_type)],\n                                                           vec![ValueDefinition\n                                                                ::\n                                                                new(\"x\",\n                                                                    RecordConstruction\n                                                                    ::\n                                                                    new(reference_type\n                                                                        .\n                                                                        clone(),\n                                                                        vec !\n                                                                        [(\"foo\"\n                                                                          .\n                                                                          into(),\n                                                                          Number\n                                                                          ::\n                                                                          new(42.0,\n                                                                              SourceInformation\n                                                                              ::\n                                                                              dummy())\n                                                                          .\n                                                                          into(),)]\n                                                                        .\n                                                                        into_iter()\n                                                                        .\n                                                                        collect(),\n                                                                        SourceInformation\n                                                                        ::\n                                                                        dummy(),),\n                                                                    reference_type,\n                                                                    SourceInformation\n                                                                    ::\n                                                                    dummy(),)\n                                                                . into()]))"
---
Ok(
    Module {
        path: ModulePath {
            package: Package {
                name: "",
                version: "",
            },
            components: [],
        },
        type_definitions: [
            TypeDefinition {
                name: "Foo",
                type_: Record(
                    Record {
                        name: "Foo",
                        elements: {
                            "foo": Number(
                                Number {
                                    source_information: SourceInformation {
                                        source_name: "",
                                        location: Location {
                                            line_number: 0,
                                            column_number: 0,
                                        },
                                        line: "",
                                    },
                                },
                            ),
                        },
                        source_information: SourceInformation {
                            source_name: "",
                            location: Location {
                                line_number: 0,
                                column_number: 0,
                            },
                            line: "",
                        },
                    },
                ),
            },
        ],
        definitions: [
            ValueDefinition(
                ValueDefinition {
                    name: "x",
                    body: RecordConstruction(
                        RecordConstruction {
                            type_: Reference(
                                Reference {
                                    name: "Foo",
                                    source_information: SourceInformation {
                                        source_name: "",
                                        location: Location {
                                            line_number: 0,
                                            column_number: 0,
                                        },
                                        line: "",
                                    },
                                },
                            ),
                            elements: {
                                "foo": Number(
                                    Number {
                                        value: 42.0,
                                        source_information: SourceInformation {
                                            source_name: "",
                                            location: Location {
                                                line_number: 0,
                                                column_number: 0,
                                            },
                                            line: "",
                                        },
                                    },
                                ),
                            },
                            source_information: SourceInformation {
                                source_name: "",
                                location: Location {
                                    line_number: 0,
                                    column_number: 0,
                                },
                                line: "",
                            },
                        },
                    ),
                    type_: Reference(
                        Reference {
                            name: "Foo",
                            source_information: SourceInformation {
                                source_name: "",
                                location: Location {
                                    line_number: 0,
                                    column_number: 0,
                                },
                                line: "",
                            },
                        },
                    ),
                    source_information: SourceInformation {
                        source_name: "",
                        location: Location {
                            line_number: 0,
                            column_number: 0,
                        },
                        line: "",
                    },
                },
            ),
        ],
        export: Export {
            names: {},
        },
        imports: [],
    },
)
